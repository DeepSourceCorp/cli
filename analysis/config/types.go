package config

import "github.com/deepsourcelabs/cli/analysis/lsp"

///////////////////////////////////////////////////////
// DSConfig is the struct for .deepsource.toml file //
/////////////////////////////////////////////////////

type DSConfig struct {
	Version         int           `mapstructure:"version,omitempty" json:"version" toml:"version"`
	ExcludePatterns []string      `mapstructure:"exclude_patterns,omitempty" json:"exclude_patterns,omitempty" toml:"exclude_patterns"`
	TestPatterns    []string      `mapstructure:"test_patterns,omitempty" json:"test_patterns,omitempty" toml:"test_patterns"`
	Analyzers       []Analyzer    `mapstructure:"analyzers,omitempty" json:"analyzers,omitempty" toml:"analyzers"`
	Transformers    []Transformer `mapstructure:"transformers,omitempty" json:"transformers,omitempty" toml:"transformers"`
}

type Analyzer struct {
	Name                string      `mapstructure:"name,omitempty" json:"name,omitempty" toml:"name"`
	RuntimeVersion      string      `mapstructure:"runtime_version,omitempty" json:"runtime_version,omitempty" toml:"runtime_version,omitempty"`
	Enabled             bool        `mapstructure:"enabled,omitempty" json:"enabled" toml:"enabled"`
	DependencyFilePaths []string    `mapstructure:"dependency_file_paths,omitempty" json:"dependency_file_paths,omitempty"`
	Meta                interface{} `mapstructure:"meta,omitempty" json:"meta,omitempty" toml:"meta"`
	Thresholds          interface{} `mapstructure:"thresholds,omitempty" json:"thresholds,omitempty"`
}

type Transformer struct {
	Name    string `mapstructure:"name,omitempty" json:"name,omitempty" toml:"name"`
	Enabled bool   `mapstructure:"enabled,omitempty" json:"enabled,omitempty" toml:"enabled"`
}

///////////////////////////////////////
// LSP based Analysis Config Types  //
/////////////////////////////////////

type AnalysisConfig struct {
	Files           []lsp.TextDocumentItem `json:"files"`
	TestFiles       []lsp.TextDocumentItem `json:"test_files"`
	ExcludedFiles   []lsp.TextDocumentItem `json:"excluded_files"`
	ExcludePatterns []string               `json:"exclude_patterns"`
	TestPatterns    []string               `json:"test_patterns"`
	AnalyzerMeta    interface{}            `json:"analyzer_meta"`
}

//////////////////////////////////////
// LSP based Analysis Result Types  //
/////////////////////////////////////

type Namespace struct {
	Key   string  `json:"key"`
	Value float64 `json:"value"`
}

type Metric struct {
	MetricCode string      `json:"metric_code"`
	Namespaces []Namespace `json:"namespaces"`
}

type AnalysisResult struct {
	Issues   []lsp.Diagnostic `json:"issues"`
	Metrics  []Metric         `json:"metrics,omitempty"`
	IsPassed bool             `json:"is_passed"`
	Errors   []Error          `json:"errors"`
	// Errors    []lsp.Diagnostic `json:"errors"`
	ExtraData interface{} `json:"extra_data"`
}

////////////////////////////////////
// Default Analysis Result Types  //
///////////////////////////////////

type Error struct {
	HMessage string `json:"hmessage"`
	Level    int    `json:"level"`
}

type Coordinate struct {
	Line   int `json:"line"`
	Column int `json:"column"`
}

type Position struct {
	Begin Coordinate `json:"begin"`
	End   Coordinate `json:"end"`
}

type Location struct {
	Path     string   `json:"path"`
	Position Position `json:"position"`
}

type Issue struct {
	Code     string   `json:"issue_code"`
	Title    string   `json:"issue_text"`
	Location Location `json:"location"`
}

type DefaultAnalysisResult struct {
	Issues    []Issue     `json:"issues"`
	Metrics   []Metric    `json:"metrics,omitempty"`
	IsPassed  bool        `json:"is_passed"`
	Errors    []Error     `json:"errors"`
	ExtraData interface{} `json:"extra_data"`
}
