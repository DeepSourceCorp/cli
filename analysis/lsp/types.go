package lsp

//////////////////////////////
// Document identity types  //
/////////////////////////////

type DocumentURI string

type TextDocumentItem struct {
	URI        DocumentURI `json:"uri"`
	LanguageID string      `json:"languageID,omitempty"`
	Version    int         `json:"version,omitempty"`
	Text       string      `json:"text,omitempty"`
}

///////////////////////////////
// Diagnostic related types //
/////////////////////////////

type DiagnosticSeverity int

const (
	Error       DiagnosticSeverity = 1
	Warning     DiagnosticSeverity = 2
	Information DiagnosticSeverity = 3
	Hint        DiagnosticSeverity = 4
)

type Position struct {
	Line      int `json:"line"`
	Character int `json:"character"`
}
type Range struct {
	Start Position `json:"start"`
	End   Position `json:"end"`
}

type Location struct {
	URI   string `json:"uri"`
	Range Range  `json:"range"`
}

type DiagnosticRelatedInformation struct {
	Location Location `json:"location"`
	Message  string   `json:"message"`
}

type Diagnostic struct {
	Range    Range              `json:"range"`
	Severity DiagnosticSeverity `json:"severity,omitempty"`
	Code     string             `json:"code,omitempty"`
	Source   string             `json:"source,omitempty"`
	Message  string             `json:"message"`

	/**
	 * An array of related diagnostic information, e.g. when symbol-names within
	 * a scope collide all definitions can be marked via this property.
	 * var a,b
	 * a := 2
	 * Issues in line 1 and 2 are related.
	 */
	RelatedInformation []DiagnosticRelatedInformation `json:"relatedInformation"`
}
