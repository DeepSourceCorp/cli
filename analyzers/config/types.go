package config

type Analysis struct {
	Command string `toml:"command" json:"command" validate:"required"`
}

// TODO: Remove the comment when we start support for Autofix.
// type Autofix struct {
//     Command string `toml:"command" json:"command" analyzertoml:"required"`
// }

type Build struct {
	Builder    string `toml:"builder" json:"builder" validate:"omitempty"`
	Dockerfile string `toml:"dockerfile" json:"dockerfile,omitempty" validate:"omitempty"`
	Script     string `toml:"script" json:"script" validate:"omitempty"`
}

type URL struct {
	Source        string `toml:"source" json:"source" validate:"omitempty,url"`
	Documentation string `toml:"documentation" json:"documentation,omitempty" validate:"omitempty,url"`
	BugTracker    string `toml:"bug_tracker" json:"bug_tracker" validate:"omitempty,url"`
}

type Test struct {
	Command string `toml:"command" json:"command" validate:"omitempty"`
}

// analyzer.toml type
type AnalyzerMetadata struct {
	// Analyzer specific data
	Name                 string            `toml:"name" json:"name" validate:"required"`
	Shortcode            string            `toml:"shortcode" json:"shortcode" validate:"required"` // New
	Description          string            `toml:"description" json:"description" validate:"required"`
	Category             string            `toml:"category" json:"category" validate:"required"`
	Tags                 []string          `toml:"tags" json:"tags,omitempty" validate:"omitempty"`
	URL                  URL               `toml:"urls" json:"urls,omitempty" validate:"omitempty"`
	EnvironmentVariables map[string]string `toml:"environment_variables" json:"environment_variables,omitempty" validate:"omitempty"`

	// Analyzer, Autofix and Transformer config
	Analysis Analysis `toml:"analysis" json:"analysis" validate:"required"`
	// Autofix Autofix `toml:"autofix" json:"autofix" validate:"omitempty"`

	// Build steps needed to build the analyzer
	Build Build `toml:"build" json:"build" validate:"omitempty"`

	// Test command for running the testing workflow
	Test Test `toml:"test" json:"test" validate:"omitempty"`
}

// Analyzer issue description
type AnalyzerIssue struct {
	Shortcode   string `validate:"omitempty"`
	Title       string `toml:"title" json:"title" validate:"required"`
	Description string `toml:"description" json:"description" validate:"required"`
	Category    string `toml:"category" json:"category" validate:"required"`
}
