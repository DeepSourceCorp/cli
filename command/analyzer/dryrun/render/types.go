package render

import (
	"html/template"
	"time"

	"github.com/deepsourcelabs/cli/types"
	"github.com/labstack/echo/v4"
)

type RunSummary struct {
	RunDuration       string // Time taken to complete analysis.
	TimeSinceRun      string // Time elapsed since the completion of the analysis run.
	AnalysisStartTime time.Time
	AnalysisEndTime   time.Time
}

type VCSInfo struct {
	Branch      string // VCS branch of the Analyzer.
	CommitSHA   string // The latest commit SHA of the Analyzer.
	VersionDiff string // The string specifying the status of Analyzer w.r.t previous version.
}

type OccurenceData struct {
	IssueMeta  types.AnalyzerIssue // Contains the data stored in issue TOMLs for the respective issue.
	Files      []string            // Files where this issue has been reported.
	FilesInfo  string              // The string containing the data of which files the issue has been reported in.
	Occurences []types.Issue       // The slice of occurences for a certain issue code.
}

type ResultData struct {
	UniqueIssuesCount     int                      // The unique issues count.
	TotalOccurences       int                      // Total issues reported by the Analyzer.
	SourcePath            string                   // The path where the source code to be analyzer is stored.
	IssuesOccurenceMap    map[string]OccurenceData // The map of issue code to its occurences data.
	IssueCategoryCountMap map[string]int           // The map of issue category to the count of the issues of that category.
	AnalysisResult        types.AnalysisResult     // The analysis result post running processors.
	MetricsMap            map[string]float64       // The map of metric names to their values.
	RenderedSourceCode    []template.HTML          // The slice containing the source code snippets for each occurence.
}

type ResultRenderOpts struct {
	EchoServer           *echo.Echo         // The Echo server instance to run the renderer server.
	Template             *template.Template // The go template field so that it can be accessible in `route.go` as well.
	PageTitle            string             // The title of the HTML page.
	AnalyzerShortcode    string             // The shortcode of the Analyzer.
	VCSInfo              VCSInfo            // The VCS information of the Analyzer.
	Summary              RunSummary         // The run summary.
	AnalysisResultData   ResultData         // The analysis result data.
	SelectedIssueCode    string             // The field used to recognize which issue code the user has clicked on to check its occurences.
	SelectedCategory     string             // The field used to recognize which category the user has clicked to filter the issues based on it.
	IssueCategoryNameMap map[string]string  // The map used to route category names to their codes. Eg: `Documentation`->`doc`.
	MetricNameMap        map[string]string  // The map of metrics shortcodes with their names.
}
