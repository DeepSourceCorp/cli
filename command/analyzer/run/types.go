package run

type AnalysisConfig struct {
	Files           []string    `json:"files"`
	ExcludePatterns []string    `json:"exclude_patterns"`
	ExcludeFiles    []string    `json:"exclude_files"`
	TestFiles       []string    `json:"test_files"`
	TestPatterns    []string    `json:"test_patterns"`
	AnalyzerMeta    interface{} `json:"analyzer_meta"`
}
