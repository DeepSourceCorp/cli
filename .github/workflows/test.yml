name: Unit Tests

on:
  push:
    branches: [develop]
  pull_request:
    branches: [develop]

jobs:
  test:
    runs-on: ubuntu-latest
    env:
      NODE_AUTH_TOKEN: ${{secrets.GITHUB_TOKEN}}
      ZEAL_SECRET: ${{secrets.GITHUB_TOKEN}}

    steps:
      - name: Checkout Bifrost
        uses: actions/checkout@v2

      - name: Setting Up Node
        uses: actions/setup-node@v2
        with:
          node-version: '14.x'
          always-auth: true
          registry-url: https://npm.pkg.github.com/
          scope: '@deepsourcelabs'
          cache: 'yarn'

      - name: Install packages
        run: yarn

      - name: Run Test
        run: yarn test --coverage=true --coverageReporters=cobertura

      - name: Install DeepSource CLI
        run: curl https://deepsource.io/cli | sh

      - name: Set DEEPSOURCE_DSN
        run: |
          ./bin/deepsource report --analyzer test-coverage --key javascript --value-file ./coverage/cobertura-coverage.xml
        env:
          DEEPSOURCE_DSN: ${{ secrets.BIFROST_DSN }}
