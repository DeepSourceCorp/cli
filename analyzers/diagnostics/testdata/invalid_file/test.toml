name = "2do-Checker"
