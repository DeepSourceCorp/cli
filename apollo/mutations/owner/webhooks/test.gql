mutation testWebhook($webhookId: ID!) {
  testWebhook(input: { webhookId: $webhookId }) {
    ok
  }
}
