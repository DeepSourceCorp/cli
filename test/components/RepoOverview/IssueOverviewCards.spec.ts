import { render } from '@testing-library/vue'
import { TooltipDirective } from 'floating-vue'

import { StatSection } from '~/components/Metrics'
import { IssueOverviewCards } from '~/components/RepoOverview'

import { storeModulesGenerator } from '~/test/mocks'
import { VcsProviderChoices } from '~/types/types'

describe('[[ IssueOverviewCards ]]', () => {
  const mocks = {
    $fetchState: { pending: false },
    $gateKeeper: {
      repo: () => ''
    },
    $socket: {
      $on: () => '',
      $off: () => ''
    }
  }

  const storeMock = {
    modules: storeModulesGenerator({
      'repository/detail': {
        namespaced: true,
        state: {
          repository: {
            id: 'UmVwb3NpdG9yeTp6dmp2eXo=',
            vcsProvider: VcsProviderChoices.Github,
            isActivated: true,
            widgets: [
              'antipattern-widget',
              'bug-risk-widget',
              'performance-widget',
              'security-widget',
              'lcv-widget',
              'bcv-widget',
              'ccv-widget',
              'cpcv-widget'
            ],
            widgetsDisplay: {
              'antipattern-widget': {
                title: 'Anti-patterns',
                value_display: 308,
                link: '/gh/deepsourcelabs/bifrost/issues/?category=antipattern',
                description: 'Antipattern issues',
                hint: null,
                has_trend_value: true,
                trend_direction: 'down',
                trend_value: '2%',
                trend_display: 'since last week',
                trend_positive: false
              },
              'bug-risk-widget': {
                title: 'Bug risks',
                value_display: 10,
                link: '/gh/deepsourcelabs/bifrost/issues/?category=bug-risk',
                description: 'Bug-Risk issues',
                hint: null,
                has_trend_value: false,
                trend_direction: null,
                trend_value: null,
                trend_display: 'no change since last week',
                trend_positive: false
              },
              'performance-widget': {
                title: 'Performance Issues',
                value_display: 27,
                link: '/gh/deepsourcelabs/bifrost/issues/?category=performance',
                description: 'Performance issues',
                hint: null,
                has_trend_value: false,
                trend_direction: null,
                trend_value: null,
                trend_display: 'no change since last week',
                trend_positive: false
              },
              'security-widget': {
                title: 'Security Issues',
                value_display: 32,
                link: '/gh/deepsourcelabs/bifrost/issues/?category=security',
                description: 'Security issues',
                hint: null,
                has_trend_value: false,
                trend_direction: null,
                trend_value: null,
                trend_display: 'no change since last week',
                trend_positive: false
              },
              'lcv-widget': {
                title: 'Line Coverage',
                value_display: '65.5%',
                link: '/gh/deepsourcelabs/bifrost/metrics/LCV',
                description: 'Code coverage of this repository excluding the branches.',
                hint: null,
                has_trend_value: false,
                trend_display: 'no change since last commit',
                trend_direction: null,
                trend_value: null,
                trend_positive: true
              },
              'bcv-widget': {
                title: 'Branch Coverage',
                value_display: '48.8%',
                link: '/gh/deepsourcelabs/bifrost/metrics/BCV',
                description:
                  'Code coverage of this repository, including the branches in the code.',
                hint: null,
                has_trend_value: false,
                trend_display: 'no change since last commit',
                trend_direction: null,
                trend_value: null,
                trend_positive: true
              },
              'ccv-widget': {
                title: 'Condition Coverage',
                value_display: '50%',
                link: '/gh/deepsourcelabs/bifrost/metrics/CCV',
                description:
                  'Percentage of logical conditions in the source code covered by tests.',
                hint: null,
                has_trend_value: false,
                trend_display: 'no change since last commit',
                trend_direction: null,
                trend_value: null,
                trend_positive: true
              },
              'cpcv-widget': {
                title: 'Composite Coverage',
                value_display: '62.5%',
                link: '/gh/deepsourcelabs/bifrost/metrics/CPCV',
                description:
                  'Weighted average of the coverage of lines, conditions and branches executed by the tests.',
                hint: null,
                has_trend_value: false,
                trend_display: 'no change since last commit',
                trend_direction: null,
                trend_value: null,
                trend_positive: true
              }
            }
          }
        }
      }
    })
  }

  const stubs = {
    CustomizeWidgetsModal: true,
    Portal: true,
    StatCard: true,
    ZIcon: true
  }

  test('renders the issue overview widgets', () => {
    const { html } = render(
      IssueOverviewCards,
      {
        components: { StatSection },
        mocks,
        store: storeMock,
        stubs
      },
      (vue) => {
        vue.directive('tooltip', TooltipDirective)
      }
    )

    expect(html()).toMatchSnapshot()
  })
})
