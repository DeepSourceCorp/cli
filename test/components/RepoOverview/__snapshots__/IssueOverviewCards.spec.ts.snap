// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[[ IssueOverviewCards ]] renders the issue overview widgets 1`] = `
"<section class=\\"flex flex-col justify-between\\">
  <!---->
  <div class=\\"flex-grow gap-2 grid grid-cols-2 lg:grid-cols-3 gap-3\\">
    <statcard-stub trendvalue=\\"2%\\" trenddirection=\\"down\\" trendhint=\\"since last week\\" trendpositive=\\"true\\" removestyles=\\"true\\" to=\\"/gh/deepsourcelabs/bifrost/issues/?category=antipattern\\" hintastooltip=\\"true\\" valuetextsize=\\"text-base\\" class=\\"lg:text-border-b-transparent group h-22 cursor-auto rounded-md bg-ink-300 p-3 text-sm font-medium hover:bg-ink-200 lg:h-26\\"> <span class=\\"text-vanilla-100 font-mono\\">
          308
        </span></statcard-stub>
    <statcard-stub trendhint=\\"no change since last week\\" trendpositive=\\"true\\" removestyles=\\"true\\" to=\\"/gh/deepsourcelabs/bifrost/issues/?category=bug-risk\\" hintastooltip=\\"true\\" valuetextsize=\\"text-base\\" class=\\"lg:text-border-b-transparent group h-22 cursor-auto rounded-md bg-ink-300 p-3 text-sm font-medium hover:bg-ink-200 lg:h-26\\"> <span class=\\"text-vanilla-100 font-mono\\">
          10
        </span></statcard-stub>
    <statcard-stub trendhint=\\"no change since last week\\" trendpositive=\\"true\\" removestyles=\\"true\\" to=\\"/gh/deepsourcelabs/bifrost/issues/?category=performance\\" hintastooltip=\\"true\\" valuetextsize=\\"text-base\\" class=\\"lg:text-border-b-transparent group h-22 cursor-auto rounded-md bg-ink-300 p-3 text-sm font-medium hover:bg-ink-200 lg:h-26\\"> <span class=\\"text-vanilla-100 font-mono\\">
          27
        </span></statcard-stub>
    <statcard-stub trendhint=\\"no change since last week\\" trendpositive=\\"true\\" removestyles=\\"true\\" to=\\"/gh/deepsourcelabs/bifrost/issues/?category=security\\" hintastooltip=\\"true\\" valuetextsize=\\"text-base\\" class=\\"lg:text-border-b-transparent group h-22 cursor-auto rounded-md bg-ink-300 p-3 text-sm font-medium hover:bg-ink-200 lg:h-26\\"> <span class=\\"text-vanilla-100 font-mono\\">
          32
        </span></statcard-stub>
    <statcard-stub trendhint=\\"no change since last commit\\" trendpositive=\\"true\\" removestyles=\\"true\\" to=\\"/gh/deepsourcelabs/bifrost/metrics/LCV\\" hintastooltip=\\"true\\" valuetextsize=\\"text-base\\" class=\\"lg:text-border-b-transparent group h-22 cursor-auto rounded-md bg-ink-300 p-3 text-sm font-medium hover:bg-ink-200 lg:h-26\\"> <span class=\\"text-vanilla-100\\">
          65.5%
        </span></statcard-stub>
    <statcard-stub trendhint=\\"no change since last commit\\" trendpositive=\\"true\\" removestyles=\\"true\\" to=\\"/gh/deepsourcelabs/bifrost/metrics/BCV\\" hintastooltip=\\"true\\" valuetextsize=\\"text-base\\" class=\\"lg:text-border-b-transparent group h-22 cursor-auto rounded-md bg-ink-300 p-3 text-sm font-medium hover:bg-ink-200 lg:h-26\\"> <span class=\\"text-vanilla-100\\">
          48.8%
        </span></statcard-stub>
    <statcard-stub trendhint=\\"no change since last commit\\" trendpositive=\\"true\\" removestyles=\\"true\\" to=\\"/gh/deepsourcelabs/bifrost/metrics/CCV\\" hintastooltip=\\"true\\" valuetextsize=\\"text-base\\" class=\\"lg:text-border-b-transparent group h-22 cursor-auto rounded-md bg-ink-300 p-3 text-sm font-medium hover:bg-ink-200 lg:h-26\\"> <span class=\\"text-vanilla-100\\">
          50%
        </span></statcard-stub>
    <statcard-stub trendhint=\\"no change since last commit\\" trendpositive=\\"true\\" removestyles=\\"true\\" to=\\"/gh/deepsourcelabs/bifrost/metrics/CPCV\\" hintastooltip=\\"true\\" valuetextsize=\\"text-base\\" class=\\"lg:text-border-b-transparent group h-22 cursor-auto rounded-md bg-ink-300 p-3 text-sm font-medium hover:bg-ink-200 lg:h-26\\"> <span class=\\"text-vanilla-100\\">
          62.5%
        </span></statcard-stub>
    <portal-stub to=\\"modal\\">
      <!---->
    </portal-stub>
  </div>
  <div></div>
</section>"
`;
